processor/batchprocessor/batch_processor_test.go

```diff
@@ -819,3 +822,168 @@ func TestShutdown(t *testing.T) {
 	factory := NewFactory()
 	processortest.VerifyShutdown(t, factory, factory.CreateDefaultConfig())
 }
+
+type metadataTracesSink struct {
+	*consumertest.TracesSink
+
+	lock               sync.Mutex
+	spanCountByToken12 map[string]int
+}
+
+func formatTwo(first, second []string) string {
+	return fmt.Sprintf("%s;%s", first, second)
+}
+
+func (mts *metadataTracesSink) ConsumeTraces(ctx context.Context, td ptrace.Traces) error {
+	info := client.FromContext(ctx)
+	token1 := info.Metadata.Get("token1")
+	token2 := info.Metadata.Get("token2")
+	mts.lock.Lock()
+	defer mts.lock.Unlock()
+
+	mts.spanCountByToken12[formatTwo(
+		token1,
+		token2,
+	)] += td.SpanCount()
+	return mts.TracesSink.ConsumeTraces(ctx, td)
+}
+
+func TestBatchProcessorSpansBatchedByMetadata(t *testing.T) {
+	sink := &metadataTracesSink{
+		TracesSink:         &consumertest.TracesSink{},
+		spanCountByToken12: map[string]int{},
+	}
+	cfg := createDefaultConfig().(*Config)
+	cfg.SendBatchSize = 1000
+	cfg.Timeout = 10 * time.Minute
+	cfg.MetadataKeys = []string{"token1", "token2"}
+	creationSet := processortest.NewNopCreateSettings()
+	creationSet.MetricsLevel = configtelemetry.LevelDetailed
+	batcher, err := newBatchTracesProcessor(creationSet, sink, cfg, false)
+	require.NoError(t, err)
+	require.NoError(t, batcher.Start(context.Background(), componenttest.NewNopHost()))
+
+	bg := context.Background()
+	callCtxs := []context.Context{
+		client.NewContext(bg, client.Info{
+			Metadata: client.NewMetadata(map[string][]string{
+				"token1": {"single"},
+				"token3": {"n/a"},
+			}),
+		}),
+		client.NewContext(bg, client.Info{
+			Metadata: client.NewMetadata(map[string][]string{
+				"token1": {"single"},
+				"token2": {"one", "two"},
+				"token4": {"n/a"},
+			}),
+		}),
+		client.NewContext(bg, client.Info{
+			Metadata: client.NewMetadata(map[string][]string{
+				"token1": nil,
+				"token2": {"single"},
+			}),
+		}),
+		client.NewContext(bg, client.Info{
+			Metadata: client.NewMetadata(map[string][]string{
+				"token1": {"one", "two", "three"},
+				"token2": {"single"},
+				"token3": {"n/a"},
+				"token4": {"n/a", "d/c"},
+			}),
+		}),
+	}
+	expectByContext := make([]int, len(callCtxs))
+
+	requestCount := 1000
+	spansPerRequest := 33
+	sentResourceSpans := ptrace.NewTraces().ResourceSpans()
+	for requestNum := 0; requestNum < requestCount; requestNum++ {
+		td := testdata.GenerateTraces(spansPerRequest)
+		spans := td.ResourceSpans().At(0).ScopeSpans().At(0).Spans()
+		for spanIndex := 0; spanIndex < spansPerRequest; spanIndex++ {
+			spans.At(spanIndex).SetName(getTestSpanName(requestNum, spanIndex))
+		}
+		td.ResourceSpans().At(0).CopyTo(sentResourceSpans.AppendEmpty())
+		// Choose a random context
```

Comment:
Is this only so that all contexts have some data? Personally, I try to make the tests predictable and reproducible, so that it gets easier to debug if it fails. If I'm reading this right, you could replace the random bits with `requestNum % len(callCtxs)`.

Reply:
Fixed.